internal/pkg/db/mongo/metadata.go

```diff
@@ -146,8 +145,13 @@ func (mc MongoClient) AddDevice(d contract.Device) (string, error) {
 	}
 
 	mapped.TimestampForAdd()
+	if err = errorMap(col.Insert(mapped)); err != nil {
+		return id, err
+	}
 
-	return id, errorMap(col.Insert(mapped))
+	//add commands based on DeviceProfile.CommandProfile
+	err = mc.addCommands(commands, mapped.Uuid, mapped.Name)
```

Comment:
It looks like if we have a problem persisting the commands in the `addCommands` function the database will be in an undesirable state since we will have already persisted the device, but not the commands. Should we add some additional protection against this?

Reply:
The point is valid but nothing in our Mongo layer supports transactions. That would be a system-wide refactoring. In this case if there were an error, then the addCommands would throw an error. That would bubble up to the router and be returned to the client. The client could then call AddDevice again and the subsequent request would return a new ID and be its own entity. There would be some manual housekeeping necessary to remove the orphaned device record, true.